core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1048,9 +1059,16 @@ class Log(@volatile var dir: File,
       lock synchronized {
         // remove the segments for lookups
         deletable.foreach(deleteSegment)
-        logStartOffset = math.max(logStartOffset, segments.firstEntry().getValue.baseOffset)
+        logStartOffset = math.max(logStartOffset, segments.firstEntry.getValue.baseOffset)
         leaderEpochCache.clearAndFlushEarliest(logStartOffset)
-        producerStateManager.evictUnretainedProducers(logStartOffset)
+        producerStateManager.truncateHead(logStartOffset)
+        updateFirstUnstableOffset()
+      }
+    } else {
+      lock synchronized {
```

Comment:
This is a question you can ignore for now since it's for my understanding: but how could the `logStartOffset` be in the middle of a segment if `numToDelete == 0` ?